src/contrib/sort/sort.cc

```diff
@@ -0,0 +1,122 @@
+/*!
+ *  Copyright (c) 2017 by Contributors
+ * \file Use standard C library call.
+ */
+
+#include <tvm/runtime/registry.h>
+#include <tvm/runtime/util.h>
+#include <dlpack/dlpack.h>
+#include <algorithm>
+#include <vector>
+
+namespace tvm {
+namespace contrib {
+
+using namespace runtime;
+
+template<typename DType>
+struct SortElem {
+  DType value;
+  int32_t index;
+
+  SortElem(DType v, int32_t i) {
+    value = v;
+    index = i;
+  }
+};
+
+template<typename DType>
+bool compare_ascend(SortElem<DType> lhs, SortElem<DType> rhs) {
+  return lhs.value < rhs.value;
+}
+
+template<typename DType>
+bool compare_descend(SortElem<DType> lhs, SortElem<DType> rhs) {
+  return lhs.value > rhs.value;
+}
+
+
+// Argsort implemented C library sort.
+// Return indices of sorted tensor.
+// By default, the last axis will be used to sort.
+// sort_num specify the number of elements to be sorted.
+// If input tensor has dimension (d1, d2, ..., d(k-1), dk, d(k+1), ..., dn)
+// and sort axis is dk. sort_num should have dimension of
+// (d1, d2, ..., d(k-1), d(k+1), ..., dn).
+TVM_REGISTER_GLOBAL("tvm.contrib.sort.argsort")
+.set_body([](TVMArgs args, TVMRetValue *ret) {
+  DLTensor *input = args[0];
+  DLTensor *sort_num = args[1];
+  DLTensor *output = args[2];
+  int32_t axis = args[3];
+  bool is_descend = args[4];
+
+  auto dtype = input->dtype;
+  auto data_ptr = static_cast<float *>(input->data);
+  auto sort_num_ptr = static_cast<int32_t *>(sort_num->data);
+  int64_t num_sort_vec = 1;
+  // Currently only supports input dtype to be float32.
+  CHECK_EQ(dtype.code, 2) << "Currently only supports input dtype "
+      "to be float32.";
+  CHECK_EQ(dtype.bits, 32) << "Currently only supports input dtype "
+      "to be float32.";
+  std::vector<SortElem<float>> sorter;
+  std::vector<int64_t> non_sort_axis;
+  if (axis < 0) {
+    axis = input->ndim + axis;
+  }
+  CHECK_LT(axis, input->ndim) << "Axis out of boundary for "
+      "input ndim " << input->ndim;
+  for (int i = 0; i < input->ndim; ++i) {
+    if (i != axis) {
+      num_sort_vec *= input->shape[i];
+      non_sort_axis.push_back(input->shape[i]);
```

Comment:
seems non_sort_axis was not necessary, we could just iterate over dim and skip the non-sort one